### ISSUE TITLE
Make `cat()` reuse file descriptors

### ISSUE THREAD

**Issue by sanmai-NL** [iovisor/bpftrace#2558]
This is not really a defect in the strict sense, it's a scalability shortcoming.

### What reproduces the defect?

Some tracing programs `cat` the same file across many probe definitions. For example, to log data read from  sysfs files. Given that I'm running out of file descriptors with a few hundred probe definitions with three `cat` statements repeated across these probe definitions, I suppose that bpftrace BPF programs opens the file for each cat call, and keeps all of them open.

Instead, I hope that bptrace can keep a global mapping of path to be opened to file descriptor, and reuse the appropriate file descriptor at each cat call. In a minimal implementation, no path normalization would be needed.

```bpftrace
BEGIN {
    cat("/sys/some_path");
    cat("/sys/some_path");
    cat("/sys/some_path");
    // ... repeated many times
}
```

### bpftrace --info

```text
System
  OS: Linux 6.1.0-0.deb11.5-amd64 #1 SMP PREEMPT_DYNAMIC Debian 6.1.12-1~bpo11+1 (2023-03-05)
  Arch: x86_64

Build
  version: v0.13.0
  LLVM: 8.0.1
  foreach_sym: yes
  unsafe uprobe: no
  bfd: yes
  bpf_attach_kfunc: yes
  bcc_usdt_addsem: yes
  bcc bpf_attach_uprobe refcount: yes
  bcc library path resolution: yes
  libbpf: yes
  libbpf btf dump: yes
  libbpf btf dump type decl: yes
  libdw (DWARF support): no

Kernel helpers
  probe_read: yes
  probe_read_str: yes
  probe_read_user: yes
  probe_read_user_str: yes
  probe_read_kernel: yes
  probe_read_kernel_str: yes
  get_current_cgroup_id: yes
  send_signal: yes
  override_return: no
  get_boot_ns: yes
  dpath: yes

Kernel features
  Instruction limit: 1000000
  Loop support: yes
  btf (depends on Build:libbpf): yes
  map batch (depends on Build:libbpf): yes
  uprobe refcount (depends on Build:bcc bpf_attach_uprobe refcount): yes

Map types
  hash: yes
  percpu hash: yes
  array: yes
  percpu array: yes
  stack_trace: yes
  perf_event_array: yes

Probe types
  kprobe: yes
  tracepoint: yes
  perf_event: yes
  kfunc: yes
  iter:task: yes
  iter:task_file: yes
```

**Comment by fbs** [comment]
are you sure the FD issue is due to cat? Each probe is an fd as well iirc, so if you have 100s of those that sounds more likely. Maybe you can up the fd limit and check `ls -hal /proc/$(pgrep -f bpftrace)/fd` or `lsof` to see if those are actally related to cat?

The cat implementation should just open the file every time:

https://github.com/iovisor/bpftrace/blob/ded5b31166219f498362672c2a81f8a5f83a522a/src/bpftrace.cpp#L580-L592

https://github.com/iovisor/bpftrace/blob/ded5b31166219f498362672c2a81f8a5f83a522a/src/utils.cpp#L924-L953

The ifile should close at the end of the function.

**Comment by sanmai-NL** [comment]
@fbs: You're right. I couldn't image the probes taking (so many) file descriptors.

I count 2119 of these file descriptors, given 210 uprobes plus 210 uretprobes.

```text
wx------ 1 root root 64 Apr  8 12:06 101 -> anon_inode:[perf_event]
lrwx------ 1 root root 64 Apr  8 12:06 1010 -> anon_inode:bpf-prog
lrwx------ 1 root root 64 Apr  8 12:06 1011 -> anon_inode:[perf_event]
lrwx------ 1 root root 64 Apr  8 12:06 1012 -> anon_inode:bpf-prog
lrwx------ 1 root root 64 Apr  8 12:06 1013 -> anon_inode:[perf_event]
lrwx------ 1 root root 64 Apr  8 12:06 1014 -> anon_inode:bpf-prog
lrwx------ 1 root root 64 Apr  8 12:06 1015 -> anon_inode:[perf_event]
lrwx------ 1 root root 64 Apr  8 12:06 1016 -> anon_inode:bpf-prog
lrwx------ 1 root root 64 Apr  8 12:06 1017 -> anon_inode:[perf_event]
lrwx------ 1 root root 64 Apr  8 12:06 1018 -> anon_inode:bpf-prog
lrwx------ 1 root root 64 Apr  8 12:06 1019 -> anon_inode:[perf_event]
...
```

How come there are so many?

**Comment by danobi** [comment]
Mind sharing the full fd list? From that truncated list it looks both reasonable and difficult to avoid: we need a FD for each prog as well as a perf FD to attach the prog.

**Comment by fbs** [comment]
yeah as soon as you insert something unique into the probe, something depending on the probe itself, it will generate a unique bpf program that needs to be loaded (1 fd) and attached (1 fd). So that goes quickly.

2k for 400 probes sounds a bit excessive tho, or are those wildcard probes?

**Comment by fbs** [comment]
also, a benefit of cat being so simple is that you avoid potential behaviour around moved files. If we open at program loading we have to check whether the file got moved or deleted or something. Now we just open, read and done

**Comment by sanmai-NL** [comment]
> also, a benefit of cat being so simple is that you avoid potentially weird behaviour around moved files. If we open at program loading we have to check whether the file got moved or deleted or something. Now we just open, read and done

When the file is (re)moved, a file descriptor kept open would still be valid. When the path is opened again, a fault would occur when that path no longer has a file.

**Comment by sanmai-NL** [comment]
> yeah as soon as you insert something unique into the probe, something depending on the probe itself, it will generate a unique bpf program that needs to be loaded (1 fd) and attached (1 fd). So that goes quickly.
> 
> 2k for 400 probes sounds a bit excessive tho, or are those wildcard probes?

I use normal uprobes and uretprobes for specific functions (all called very infrequently). The bpftrace code is generated, and I need to log which probe was triggered, which results in slightly different function bodies. The `cat()` statements and other printing statements, all that is left in the probe bodies, are constant though.

```bpftrace
 printf("['%s','%s','%s',%u,", "uprobe", "module", "tests::myfunction", nsecs);
```

**Comment by sanmai-NL** [comment]
> Mind sharing the full fd list? From that truncated list it looks both reasonable and difficult to avoid: we need a FD for each prog as well as a perf FD to attach the prog.

For my understanding: why the full list of file descriptors? Wouldn't the trace source code be more informative?

**Comment by danobi** [comment]
> > also, a benefit of cat being so simple is that you avoid potentially weird behaviour around moved files. If we open at program loading we have to check whether the file got moved or deleted or something. Now we just open, read and done
> 
> When the file is (re)moved, a file descriptor kept open would still be valid. When the path is opened again, a fault would occur when that path no longer has a file.

That is true. There are really no great options to be honest. This is one of the consequences of async actions. We should probably document this behavior.

>> Mind sharing the full fd list? From that truncated list it looks both reasonable and difficult to avoid: we need a FD for each prog as well as a perf FD to attach the prog.
>
> For my understanding: why the full list of file descriptors? Wouldn't the trace source code be more informative?

Yeah, either would work. I assumed the source had something sensitive given you have not shared it.

**Comment by fbs** [comment]
> > also, a benefit of cat being so simple is that you avoid potentially weird behaviour around moved files. If we open at program loading we have to check whether the file got moved or deleted or something. Now we just open, read and done
> 
> When the file is (re)moved, a file descriptor kept open would still be valid. When the path is opened again, a fault would occur when that path no longer has a file.

yeah that is what I meant. `cat` is now just the file as it is at the moment of calling `cat`. If you keep the FD open you have to keep track of renames and how to handle that, re-open or show the old file
